pkg/destroy/azure/azure.go

```diff
@@ -101,25 +102,85 @@ func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.
 
 // Run is the entrypoint to start the uninstall process.
 func (o *ClusterUninstaller) Run() error {
+	var errs []error
+	var err error
+
 	o.configureClients()
 
-	o.Logger.Debug("deleting public records")
-	if err := deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, o.ResourceGroupName); err != nil {
+	timeout := 30 * time.Minute
+	waitCtx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	wait.UntilWithContext(
+		waitCtx,
+		func(ctx context.Context) {
+			o.Logger.Debugf("deleting public records")
+			err = deletePublicRecords(ctx, o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, o.ResourceGroupName)
+			if err != nil {
+				o.Logger.Debug(err)
+				return
+			}
+			cancel()
+		},
+		1*time.Second,
+	)
+	err = waitCtx.Err()
+	if err != nil && err != context.Canceled {
+		errs = append(errs, errors.Wrap(err, "failed to delete public DNS records"))
 		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete public DNS records")
 	}
-	o.Logger.Debug("deleting resource group")
-	if err := deleteResourceGroup(context.TODO(), o.resourceGroupsClient, o.Logger, o.ResourceGroupName); err != nil {
+
+	deadline, _ := waitCtx.Deadline()
+	diff := deadline.Sub(time.Now())
+	if diff > 0 {
+		waitCtx, cancel = context.WithTimeout(context.Background(), diff)
+	}
```

Comment:
I don't think we need to create a new context, the previous context should work fine and will be valid for the remaining timeout..

Reply:
Ah i see Wait runs the fn until the context completes..

Reply:
It isn't though. wait.UntilWithContext() will keep calling func() until it times out or is cancelled. When the func() does not fail, cancel() is called, which invalidates the context. I can confirm this. Am I misunderstanding ?